File: 039 Combination Sum/combination_sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> entry;
        helper(candidates, target, entry, 0, result);
        return result;
    }
    
    void helper(vector<int>& c, int t, vector<int>& entry, int p, vector<vector<int>>& result) {
        if (t == 0) {
            result.push_back(entry);
            return;
        }
        for (int i = p; i < c.size(); i++) {
            if (c[i] > t) continue;
            if (i != p && c[i] == c[i - 1]) continue;
            entry.push_back(c[i]);
            helper(c, t - c[i], entry, i, result);
            entry.pop_back();
        }
    }
}; // Author: XC

File: 039 Combination Sum/combination_sum.cs
public class Solution {
    private IList<IList<int>> result;
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        Array.Sort(candidates);
        IList<int> entry = new List<int>();
        result = new List<IList<int>>();
        helper(candidates, entry, target, 0, 0);
        return result;
        
    }
    
    private void helper(int[] candidates, IList<int> entry, int target, int sum, int curr) {
        if (sum > target) return;
        if (sum == target) {
            result.Add(new List<int>(entry));
            return;
        }
        for (int i = curr; i < candidates.Length; i++) {
            entry.Add(candidates[i]);
            helper(candidates, entry, target, sum + candidates[i], i);
            entry.RemoveAt(entry.Count - 1);
        }
    }
} // Author: XC
